Support tagstack for lsp location jump

yen3 opened the pull request:
Thanks for the amazing built-in lsp feature. I really love it.

The function to jump to the definition in lsp is very good but can not jump to the original location. The patch tries to provide a simple solution - add the destination to a temp tag file then jump.

Maybe it's not a good solution. I hope that the commit can provide some ideas to let someone add the function.

The code rewrites the code from https://github.com/davidhalter/jedi-vim/blob/686f39ac002ca31acd663ae8749982144484ff95/pythonx/jedi_vim.py#L387-L405

yen3 committed:
Support tagstack for lsp location jump

The patch supports to record the current location to tagstack. The
code copies from https://github.com/davidhalter/jedi-vim/blob/686f39ac002ca31acd663ae8749982144484ff95/pythonx/jedi_vim.py#L387-L405

vheon commented:
> The function to jump to the definition in lsp is very good but can not jump to the original location. 

What do you mean? Since there is this line:

https://github.com/neovim/neovim/blob/1f56f9a4b32c7d7aebafea226c148e9ed8bbabdb/runtime/lua/vim/lsp/util.lua#L327

you should be able to go back to the previous position with `<C-o>`

yen3 commented:
Oh. I don't know <C-O> can go back to the previous location. Because the official documents recommands to map <C-]> to jump to the definition. I think <C-T> should work to jump back.

It can resolve the problem very well. Is it possible to update tagstack in a proper way?

vheon commented:
The [jumplist](https://neovim.io/doc/user/motion.html#jump-motions) and the [tagstack](https://neovim.io/doc/user/tagsrch.html#tagstack) are two distinct thing. I was only commenting on the fact that since we're adding the position in the jumplist before going to the definition you can come back with the jumplist commands. I also see a TODO the line below https://github.com/neovim/neovim/blob/1f56f9a4b32c7d7aebafea226c148e9ed8bbabdb/runtime/lua/vim/lsp/util.lua#L328 saying to update the tagstack as well; but rather than creating a temp file and reading it I would simply use the [settagstack](https://neovim.io/doc/user/eval.html#settagstack()) function to update it.

yen3 commented:
Sorry, I don't have enough knowledge to implement it. Thanks for your patient and explanation. 

In the beginning, I try to update tagstack with [settagstack](https://neovim.io/doc/user/eval.html#settagstack()), but there is one problem. We can only add item(s) to the stack and have no chance to pop out the stack in the function. Since the pop-stack behavior is defined in < C -T >. I am not familiar with neovim source code so I try to implement it in a hack way.

I will try to trace source code to find the definition of < C -T > behavior. If there is no good way to push/pop the tagstack. I will close the PR.

Thank you!

vheon commented:
> We can only add item(s) to the stack and have no chance to pop out the stack in the function.

I'm not sure why you should pop tags from the stack without the user asking :/ I mean from what it looks like we want to add the position to the tagstack before jumping so that if a user mapped `<C-]>` to the jump to definition functionality and wanted to use `<C-T>` to go back to the original position.

If however I'm missing something and the need of poping tags from the tagstack is true then I think you could simply use the [gettagstack](https://neovim.io/doc/user/eval.html#gettagstack()) t o get them all, remove what is needed to be removed and replace the whole tagstack with the new one with the usual [settagstack](https://neovim.io/doc/user/eval.html#settagstack()) which looks like can add a single tag or replace the tagstack with a new one.

yen3 commented:
Thanks for your discussion. After the initial study for the behaviour of
tagstack.  There is a need to manipulate the tagstack with `gettagstack` and
`settagstack`. Thanks for your suggestion and information.

For example, the user jumps to another location with `<C-]>` and jumps back
with `<C-T>`

| Num | Action              | Stack                 | Current Index |
| --- | ------------------- | --------------------- | ------------- |
| 1   | jump to A1          |  A1                   | 1             |
| 2   | jump to B1          |  A1 -> B1             | 2             |
| 3   | jump to C1          |  A1 -> B1 -> C1       | 3             |
| 4   | jump to D1          |  A1 -> B1 -> C1 -> D1 | 4             |
| 5   | jump back to C1     |  A1 -> B1 -> C1 -> D1 | 3             |
| 6   | jump back to B1     |  A1 -> B1 -> C1 -> D1 | 2             |
| 7   | jump back to A1     |  A1 -> B1 -> C1 -> D1 | 1             |
| 8   | jump to B2          |  A1 -> B2             | 2             |

In num7 to num8 action, the stack content has to be manipulated by current
index then append new stack item.

I will try to implement the function with `gettagstack` and `settagstack`. It
needs time to implement it since I am unfamiliar with lua and how to manipulate
vim data structure in lua.

Look forward to commit and discuss with you again.

yen3 committed:
Update tagstack with settagstack

The patch push the current location to tagstack before jump. If users
have need to jump back. They can use <C-o> (with jumplist) or <C-t>
(with tagstack) to jump back.

yen3 commented:
@vheon I commit a new implementation with `settagstack`. Please review it when you are available. Thanks for your time.

tjdevries commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -325,7 +325,13 @@ function M.jump_to_location(location)
+  -- Push a new item into tagstack
+  local items = {}
+  table.insert(items, {tagname=vim.fn.expand("<cword>"), from=vim.fn.getpos('.')})
+  vim.fn.settagstack(vim.fn.bufnr('%'), {items=items}, 't')
```
I'm not super familiar with the function, but why would you use `t` here as the argument. I thought `a` would be better for this case (since we're just adding a new thing to the stack).

Also, can you add a test for this? For example, call this and then check that the tag stack is updated in the tests?

justinmk replied:
agreed. @yen3 have you been using this for awhile and it's working as expected? Can you try with 'a'?  Then we can merge.

Adding a test would be nice but shouldn't block this.

yen3 replied:
I prefer to use `t` rather then `a` since it's possible to pop the tagstack then push again.

For example, `jump to` means `Ctrl-]` and `jump back to` means `Ctrl-t`

| Num | Action              | Stack                 | Current Index |
| --- | ------------------- | --------------------- | ------------- |
| 1   | jump to A1          |  A1                   | 1             |
| 2   | jump to B1          |  A1 -> B1             | 2             |
| 3   | jump to C1          |  A1 -> B1 -> C1       | 3             |
| 4   | jump to D1          |  A1 -> B1 -> C1 -> D1 | 4             |
| 5   | jump back to C1     |  A1 -> B1 -> C1 -> D1 | 3             |
| 6   | jump back to B1     |  A1 -> B1 -> C1 -> D1 | 2             |
| 7   | jump back to A1     |  A1 -> B1 -> C1 -> D1 | 1             |
| 8   | jump to B2          |  A1 -> B2             | 2             |

In num-7 to num-8 action, the stack content has to be manipulated by the current index then append new stack item.

If we use `a` to append in num-8, the stack is `A1 -> B1 -> C1 -> D1 -> B2` not `A1 -> B2`. The stack behaviour is not the same as original.

I have no idea how to add test case(s) for the project and run them. Is there any documentation to help to write test cases? If no, I will study it.

Thanks for your comment and discussion. @tjdevries  @justinmk

justinmk replied:
> I have no idea how to add test case(s) for the project and run them. Is there any documentation to help to write test cases? If no, I will study it.

See `test/functional/example_spec.lua` for example. You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/example_spec.lua make functionaltest

More docs at https://github.com/neovim/neovim/blob/master/test/README.md

justinmk commented:
Let's try this, It's better than the current state.

We do need to improve our test coverage of LSP, see https://github.com/neovim/neovim/pull/12096#discussion_r416671756

yen3 commented:
Thanks for your help and introduction to writing a test.

I will study it and try to send another patch in the next few weeks.
